Split std.algorithm into submodules

quickfur opened the pull request:
It's about time we split up this monster into more manageable chunks. I have managed to get the resulting pieces to compile and pass unittests... due to the high possibility of merge conflicts, it would be nice if this was merged ASAP.

The new submodules basically follow the categorization given in the original std/algorithm.d ddoc header:
* `std.algorithm.searching`
* `std.algorithm.comparison`
* `std.algorithm.iteration`
* `std.algorithm.sorting`
* `std.algorithm.setops`
* `std.algorithm.mutation`.

There will also be a corresponding dlang.org PR for updating the website navigation to handle the new submodules.

There are still some outstanding issues, like a few functions left in package.d that I've no idea where they belong. Please suggest.

Also, the ddoc headers need to be updated; I don't have any more time to work on this tonight; but since this is a large changeset, I thought I should get the review process going first. Doc commits will be forthcoming in a day or two.

andralex commented:
I've done some splitting once and they all came like one saying large files are not a problem for them. Now I don't know what to say.

quickfur commented:
The fact of the matter is that the current std/algorithm.d is so large that I can no longer run its unittests on my machine without running out of RAM. That significantly increases the amount of effort it takes to contribute to it, since I either have to submit premature PRs and let the autotester bear the brunt of the load, or I have to (hackishly) patch std/algorithm.d to delete large swaths of unrelated code just so unittests will run on my machine.

(I actually did this, btw, when doing the `groupBy` PR -- I had a diff that delete 20% of the functions in std/algorithm.d, and did a `\cp std/algorithm.d /tmp/algorithm.d && patch -p2 < /tmp/algo.diff && dmd -unittest -main -of/tmp/test /tmp/algorithm.d && /tmp/test`. Actually running the unittests via the makefiles? Not even a remote chance.)

Also, having so many unrelated (or barely-related) functions thrown into the same file means that importing std.algorithm will automatically pull in about half of Phobos due to dependent imports, even if these aren't actually necessary for the one function that you're trying to use. After splitting things up, I've managed to reduce quite a few module-level imports (mostly caused by compiler bugs/limitations, such as private imports not supporting per-symbol importing), which should significantly simplify the module dependency graph in Phobos.

JakobOvrum commented:
> I've done some splitting once and they all came like one saying large files are not a problem for them. Now I don't know what to say.

We already split std.container and std.range for 2.067, and IIRC @jmdavis has said that he's working on, or at least has worked on, trying to split std.datetime. Splitting modules seems to be quite popular as of late and was a big part of the rationale for the package.d feature - how is std.algorithm any different?

quickfur commented:
Alright, the documentation is now updated. The table of functions at the top of the module now links to the correct submodule and symbol. The cheat sheet, which was growing way too long before, has been split up and moved into the respective submodules. It's looking pretty good! (I'll see if I can get a preview site up... the recent JS dependency introduced into dlang.org has made the output less than pleasing on the eye when the JS files aren't installed into the webserver root, which is an annoyance when I'm using the root for something else... but anyway. All things are hackable if one puts his mind to it...)

quickfur commented:
Ah, turns out it was a bug that got fixed in dlang.org git HEAD. Now the preview is working and lookin' pretty: http://eusebeia.dyndns.org/~hsteoh/tmp/web/phobos-prerelease/std_algorithm_package

quickfur commented:
Merged latest changes from master... whoa, rebasing was a total nightmare because git doesn't understand the concept of moving content between files. The conflict diffs were totally screwed up and unresolvable. The only way I could make it work was to do a merge from master and apply the diff from PR #2876 by hand.

Can we pretty please merge this ASAP? Every further change to std/algorithm.d will cost me hours to merge back. :-(  (And I honestly do *not* want to attempt splitting std/algorithm.d a second time from the latest HEAD -- that in itself is already a huge amount of work.)

rikkimax commented:
I have to say, there is an awfully a lot of code in the package file. Makes me a little uncomfortable.

quickfur commented:
Yeah, those are bits that I haven't quite figured out what to do with just yet. For the most part, they should not affect everyday usage of std.algorithm (sub)modules, but eventually they should be cleaned up. I was hoping to do that after the main work (this PR) is merged. There are some common internal testing stuff that really should be factored out into a private submodule, and some stale code that probably should be deleted, as well as undocumented stuff that we need to decide what to do with. It might be a bit too much to handle for this PR, that's why I'm postponing it for afterwards, once people decide that this is the direction we want to go in, instead of spending lots of time on these little details only to have the whole thing rejected in the end.

rikkimax commented:
@quickfur Normally, for any code I have that is more or less dependent upon the rest of package, I put it into a defs module. So that package module is only for public imports and module constructors.
However, you have plans, so that's fine IMO.

quickfur commented:
Woohoo, the autotester is green now! Finally fixed those annoying windows makefile failures. Can we merge now? Pretty please?

quickfur commented:
Now we need to also pull https://github.com/D-Programming-Language/dlang.org/pull/804 to update the dlang.org navbar.

quickfur commented:
Ah, right. Thanks for merging this PR at least, though!

jacob-carlborg commented:
@quickfur the cheat sheet looks broken, the tables are gone: http://dlang.org/phobos-prerelease/std_algorithm_comparison.html

quickfur commented:
Strange, I can't seem to reproduce this problem locally! I checked out the latest phobos/dlang.org and built everything from a clean slate (even nuked my web directory to make sure there's nothing leftover from older installations), and I don't see the problem.

Maybe it's something that has been fixed in git but not yet pushed to the actual website? ping @andralex

jacob-carlborg commented:
Hmm, not that I can see for std.algorithm.comparison, perhaps something in the dlang.org repository.
